
project frameworks/base/
diff --git a/core/java/android/app/AppOpsManager.java b/core/java/android/app/AppOpsManager.java
index 11a5ee85f08..2d82d55cecc 100644
--- a/core/java/android/app/AppOpsManager.java
+++ b/core/java/android/app/AppOpsManager.java
@@ -267,7 +267,9 @@ public class AppOpsManager {
     /** @hide */
     public static final int OP_SU = 68;
     /** @hide */
-    public static final int _NUM_OP = 69;
+    public static final int OP_MOTION_SENSORS = 69;
+    /** @hide */
+    public static final int _NUM_OP = 70;
 
     /** Access to coarse location information. */
     public static final String OPSTR_COARSE_LOCATION = "android:coarse_location";
@@ -378,6 +380,8 @@ public class AppOpsManager {
             "android:data_connect_change";
     private static final String OPSTR_SU =
             "android:su";
+    private static final String OPSTR_MOTION_SENSORS =
+            "android:motion_sensors";
 
     /**
      * This maps each operation to the operation that serves as the
@@ -456,7 +460,8 @@ public class AppOpsManager {
             OP_BOOT_COMPLETED,
             OP_NFC_CHANGE,
             OP_DATA_CONNECT_CHANGE,
-            OP_SU
+            OP_SU,
+            OP_MOTION_SENSORS
     };
 
     /**
@@ -533,6 +538,7 @@ public class AppOpsManager {
             OPSTR_NFC_CHANGE,
             OPSTR_DATA_CONNECT_CHANGE,
             OPSTR_SU,
+            OPSTR_MOTION_SENSORS,
     };
 
     /**
@@ -609,6 +615,7 @@ public class AppOpsManager {
             "NFC_CHANGE",
             "DATA_CONNECT_CHANGE",
             "SU",
+            "MOTION_SENSORS",
     };
 
     /**
@@ -685,6 +692,7 @@ public class AppOpsManager {
             Manifest.permission.NFC,
             Manifest.permission.MODIFY_PHONE_STATE,
             null,
+            null,
     };
 
     /**
@@ -762,6 +770,7 @@ public class AppOpsManager {
             null, //NFC_CHANGE
             null, //DATA_CONNECT_CHANGE
             UserManager.DISALLOW_SU, //SU TODO: this should really be investigated.
+            null, //MOTION_SENSORS
     };
 
     /**
@@ -838,6 +847,7 @@ public class AppOpsManager {
             true, // NFC_CHANGE
             true, //DATA_CONNECT_CHANGE
             false, //SU
+            false, //MOTION_SENSORS
     };
 
     /**
@@ -913,6 +923,7 @@ public class AppOpsManager {
             AppOpsManager.MODE_ALLOWED, // OP_NFC_CHANGE
             AppOpsManager.MODE_ALLOWED,
             AppOpsManager.MODE_ASK, // OP_SU
+            AppOpsManager.MODE_ASK, // OP_MOTION_SENSORS
     };
 
     /**
@@ -989,6 +1000,7 @@ public class AppOpsManager {
             AppOpsManager.MODE_ASK,     // OP_NFC_CHANGE
             AppOpsManager.MODE_ASK,     // OP_DATA_CONNECT_CHANGE
             AppOpsManager.MODE_ASK,     // OP_SU
+            AppOpsManager.MODE_ASK,     // OP_MOTION_SENSORS
     };
 
     /**
@@ -1064,6 +1076,7 @@ public class AppOpsManager {
         true,     // OP_NFC_CHANGE
         true,     // OP_DATA_CONNECT_CHANGE
         true,     // OP_SU
+        true,     // OP_MOTION_SENSORS
     };
 
     /**
@@ -1143,6 +1156,7 @@ public class AppOpsManager {
             false,     // OP_NFC_CHANGE
             false,     // OP_DATA_CONNECT_CHANGE
             false,     // OP_SU
+            false,     // OP_MOTION_SENSORS
     };
 
     /**
diff --git a/core/res/res/values-de/cm_strings.xml b/core/res/res/values-de/cm_strings.xml
index 5209a36af4d..ac7323fd26b 100644
--- a/core/res/res/values-de/cm_strings.xml
+++ b/core/res/res/values-de/cm_strings.xml
@@ -166,6 +166,7 @@
   <string name="app_ops_get_accounts">Geräte-Konten abzurufen</string>
   <string name="app_ops_wifi_change">WLAN-Status zu ändern</string>
   <string name="app_ops_su">Root-Zugriff zu erhalten</string>
+  <string name="app_ops_motion_sensors">Bewegungssensoren zu nutzen</string>
   <!-- Notify user that they are in Lock-to-app (for devices without navbar) -->
   <string name="lock_to_app_toast_no_navbar">Zum Lösen dieser Ansicht drücken und halten Sie die Zurück-Taste.</string>
   <!-- Template for showing cellular network operator name while LTE calling is enabled -->
diff --git a/core/res/res/values-fr/cm_strings.xml b/core/res/res/values-fr/cm_strings.xml
index a4be3cff3f1..259bfd0536e 100644
--- a/core/res/res/values-fr/cm_strings.xml
+++ b/core/res/res/values-fr/cm_strings.xml
@@ -166,6 +166,7 @@
   <string name="app_ops_get_accounts">obtenir les comptes de l\'appareil</string>
   <string name="app_ops_wifi_change">changer l\'état du Wi-Fi</string>
   <string name="app_ops_su">obtenir l\'accès root</string>
+  <string name="app_ops_motion_sensors">utiliser les capteurs de mouvement</string>
   <!-- Notify user that they are in Lock-to-app (for devices without navbar) -->
   <string name="lock_to_app_toast_no_navbar">Pour déverrouiller l\'écran, appuyez et maintenez le bouton Retour.</string>
   <!-- Template for showing cellular network operator name while LTE calling is enabled -->
diff --git a/core/res/res/values/cm_arrays.xml b/core/res/res/values/cm_arrays.xml
index f66cb0626f7..5a20fe8282e 100644
--- a/core/res/res/values/cm_arrays.xml
+++ b/core/res/res/values/cm_arrays.xml
@@ -186,6 +186,8 @@
         <item>@string/app_ops_toggle_mobile_data</item>
         <!-- OP_SU -->
         <item>@string/app_ops_su</item>
+        <!-- OP_MOTION_SENSORS -->
+        <item>@string/app_ops_motion_sensors</item>
     </string-array>
 
     <!-- A list of pre-installed applications that will be treated as carrier apps,
diff --git a/core/res/res/values/cm_strings.xml b/core/res/res/values/cm_strings.xml
index a8fcbe52024..c8dab1b6832 100644
--- a/core/res/res/values/cm_strings.xml
+++ b/core/res/res/values/cm_strings.xml
@@ -188,6 +188,7 @@
     <string name="app_ops_get_accounts">get device accounts</string>
     <string name="app_ops_wifi_change">change Wi-Fi state</string>
     <string name="app_ops_su">get root access</string>
+    <string name="app_ops_motion_sensors">use the motion sensors</string>
 
     <!-- Notify user that they are in Lock-to-app (for devices without navbar) -->
     <string name="lock_to_app_toast_no_navbar">To unpin this screen, touch and hold the Back button.</string>
