
project frameworks/base/
diff --git a/core/java/android/app/AppOpsManager.java b/core/java/android/app/AppOpsManager.java
index 11a5ee85f08..c1679379a45 100644
--- a/core/java/android/app/AppOpsManager.java
+++ b/core/java/android/app/AppOpsManager.java
@@ -265,9 +265,13 @@ public class AppOpsManager {
     /** @hide */
     public static final int OP_DATA_CONNECT_CHANGE = 67;
     /** @hide */
-    public static final int OP_SU = 68;
+    public static final int OP_MOTION_SENSORS = 68;
     /** @hide */
-    public static final int _NUM_OP = 69;
+    public static final int OP_OTHER_SENSORS = 69;
+    /** @hide */
+    public static final int OP_SU = 70;
+    /** @hide */
+    public static final int _NUM_OP = 71;
 
     /** Access to coarse location information. */
     public static final String OPSTR_COARSE_LOCATION = "android:coarse_location";
@@ -376,6 +380,10 @@ public class AppOpsManager {
             "android:nfc_change";
     private static final String OPSTR_DATA_CONNECT_CHANGE =
             "android:data_connect_change";
+    private static final String OPSTR_MOTION_SENSORS =
+            "android:motion_sensors";
+    private static final String OPSTR_OTHER_SENSORS =
+            "android:other_sensors";
     private static final String OPSTR_SU =
             "android:su";
 
@@ -456,6 +464,8 @@ public class AppOpsManager {
             OP_BOOT_COMPLETED,
             OP_NFC_CHANGE,
             OP_DATA_CONNECT_CHANGE,
+            OP_MOTION_SENSORS,
+            OP_OTHER_SENSORS,
             OP_SU
     };
 
@@ -532,6 +542,8 @@ public class AppOpsManager {
             OPSTR_BOOT_COMPLETED,
             OPSTR_NFC_CHANGE,
             OPSTR_DATA_CONNECT_CHANGE,
+            OPSTR_MOTION_SENSORS,
+            OPSTR_OTHER_SENSORS,
             OPSTR_SU,
     };
 
@@ -608,6 +620,8 @@ public class AppOpsManager {
             "BOOT_COMPLETED",
             "NFC_CHANGE",
             "DATA_CONNECT_CHANGE",
+            "MOTION_SENSORS",
+            "OTHER_SENSORS",
             "SU",
     };
 
@@ -685,6 +699,8 @@ public class AppOpsManager {
             Manifest.permission.NFC,
             Manifest.permission.MODIFY_PHONE_STATE,
             null,
+            null,
+            null,
     };
 
     /**
@@ -761,6 +777,8 @@ public class AppOpsManager {
             null, //BOOT_COMPLETED
             null, //NFC_CHANGE
             null, //DATA_CONNECT_CHANGE
+            null, //MOTION_SENSORS
+            null, //OTHER_SENSORS
             UserManager.DISALLOW_SU, //SU TODO: this should really be investigated.
     };
 
@@ -837,6 +855,8 @@ public class AppOpsManager {
             true, // BOOT_COMPLETED
             true, // NFC_CHANGE
             true, //DATA_CONNECT_CHANGE
+            false, //MOTION_SENSORS
+            false, //OTHER_SENSORS
             false, //SU
     };
 
@@ -912,7 +932,9 @@ public class AppOpsManager {
             AppOpsManager.MODE_ALLOWED, // OP_BOOT_COMPLETED
             AppOpsManager.MODE_ALLOWED, // OP_NFC_CHANGE
             AppOpsManager.MODE_ALLOWED,
-            AppOpsManager.MODE_ASK, // OP_SU
+            AppOpsManager.MODE_ASK,     // OP_MOTION_SENSORS
+            AppOpsManager.MODE_ALLOWED, // OP_OTHER_SENSORS
+            AppOpsManager.MODE_ASK,     // OP_SU
     };
 
     /**
@@ -988,6 +1010,8 @@ public class AppOpsManager {
             AppOpsManager.MODE_ALLOWED, // OP_BOOT_COMPLETED
             AppOpsManager.MODE_ASK,     // OP_NFC_CHANGE
             AppOpsManager.MODE_ASK,     // OP_DATA_CONNECT_CHANGE
+            AppOpsManager.MODE_ASK,     // OP_MOTION_SENSORS
+            AppOpsManager.MODE_ALLOWED, // OP_OTHER_SENSORS
             AppOpsManager.MODE_ASK,     // OP_SU
     };
 
@@ -1063,6 +1087,8 @@ public class AppOpsManager {
         false,    // OP_BOOT_COMPLETED
         true,     // OP_NFC_CHANGE
         true,     // OP_DATA_CONNECT_CHANGE
+        true,     // OP_MOTION_SENSORS
+        false,    // OP_OTHER_SENSORS
         true,     // OP_SU
     };
 
@@ -1142,6 +1168,8 @@ public class AppOpsManager {
             false,     // OP_BOOT_COMPLETED
             false,     // OP_NFC_CHANGE
             false,     // OP_DATA_CONNECT_CHANGE
+            false,     // OP_MOTION_SENSORS
+            false,     // OP_OTHER_SENSORS
             false,     // OP_SU
     };
 
diff --git a/core/res/res/values-de/cm_strings.xml b/core/res/res/values-de/cm_strings.xml
index 5209a36af4d..83f0d025d46 100644
--- a/core/res/res/values-de/cm_strings.xml
+++ b/core/res/res/values-de/cm_strings.xml
@@ -165,7 +165,10 @@
   <string name="app_ops_turn_on_screen">den Bildschirm einzuschalten</string>
   <string name="app_ops_get_accounts">Geräte-Konten abzurufen</string>
   <string name="app_ops_wifi_change">WLAN-Status zu ändern</string>
+  <string name="app_ops_motion_sensors">Bewegungssensoren zu nutzen</string>
+  <string name="app_ops_other_sensors">sonstige Sensoren zu nutzen</string>
   <string name="app_ops_su">Root-Zugriff zu erhalten</string>
+
   <!-- Notify user that they are in Lock-to-app (for devices without navbar) -->
   <string name="lock_to_app_toast_no_navbar">Zum Lösen dieser Ansicht drücken und halten Sie die Zurück-Taste.</string>
   <!-- Template for showing cellular network operator name while LTE calling is enabled -->
diff --git a/core/res/res/values-fr/cm_strings.xml b/core/res/res/values-fr/cm_strings.xml
index a4be3cff3f1..e271ba7376d 100644
--- a/core/res/res/values-fr/cm_strings.xml
+++ b/core/res/res/values-fr/cm_strings.xml
@@ -165,6 +165,8 @@
   <string name="app_ops_turn_on_screen">allumer l\'écran</string>
   <string name="app_ops_get_accounts">obtenir les comptes de l\'appareil</string>
   <string name="app_ops_wifi_change">changer l\'état du Wi-Fi</string>
+  <string name="app_ops_motion_sensors">utiliser les capteurs de mouvement</string>
+  <string name="app_ops_other_sensors">utiliser d\'autres capteurs</string>
   <string name="app_ops_su">obtenir l\'accès root</string>
   <!-- Notify user that they are in Lock-to-app (for devices without navbar) -->
   <string name="lock_to_app_toast_no_navbar">Pour déverrouiller l\'écran, appuyez et maintenez le bouton Retour.</string>
diff --git a/core/res/res/values/cm_arrays.xml b/core/res/res/values/cm_arrays.xml
index f66cb0626f7..23a252d7e41 100644
--- a/core/res/res/values/cm_arrays.xml
+++ b/core/res/res/values/cm_arrays.xml
@@ -184,6 +184,10 @@
         <item>@string/app_ops_toggle_nfc</item>
         <!-- OP_DATA_CONNECT_CHANGE -->
         <item>@string/app_ops_toggle_mobile_data</item>
+        <!-- OP_MOTION_SENSORS -->
+        <item>@string/app_ops_motion_sensors</item>
+        <!-- OP_OTHER_SENSORS -->
+        <item>@string/app_ops_other_sensors</item>
         <!-- OP_SU -->
         <item>@string/app_ops_su</item>
     </string-array>
diff --git a/core/res/res/values/cm_strings.xml b/core/res/res/values/cm_strings.xml
index a8fcbe52024..e0a639515ce 100644
--- a/core/res/res/values/cm_strings.xml
+++ b/core/res/res/values/cm_strings.xml
@@ -187,6 +187,8 @@
     <string name="app_ops_turn_on_screen">turn the screen on</string>
     <string name="app_ops_get_accounts">get device accounts</string>
     <string name="app_ops_wifi_change">change Wi-Fi state</string>
+    <string name="app_ops_motion_sensors">use the motion sensors</string>
+    <string name="app_ops_other_sensors">use other sensors</string>
     <string name="app_ops_su">get root access</string>
 
     <!-- Notify user that they are in Lock-to-app (for devices without navbar) -->
